File: Source/code/forms/formSaveRange.pas
// ===================================================================
//
// (c) Paul Alan Freshney 2012-2022
// www.freshney.org :: paul@freshney.org :: maximumoctopus.com
//
// https://github.com/MaximumOctopus/LEDMatrixStudio
//
// Please do not modifiy this comment section
//
// ===================================================================


unit formSaveRange;


interface


uses
  Winapi.Windows, Winapi.Messages, System.SysUtils, System.Variants, System.Classes, Vcl.Graphics,
  Vcl.Controls, Vcl.Forms, Vcl.Dialogs, Vcl.StdCtrls, Vcl.Buttons, Vcl.ExtCtrls,

  languagehandler;


type
  TfrmSaveRange = class(TForm)
    Bevel1: TBevel;
    bOK: TBitBtn;
    bCancel: TBitBtn;
    Image1: TImage;
    GroupBox1: TGroupBox;
    lStart: TLabel;
    lEnd: TLabel;
    eStartFrame: TEdit;
    eEndFrame: TEdit;
    procedure eStartFrameChange(Sender: TObject);
    procedure FormCreate(Sender: TObject);
  private
    FFrameCount : integer;

    procedure SetGUILanguageText;

    function ValidateInputs: boolean;
  end;


  TSaveFrameRangeObject = record
    Process : boolean;
    StartFrame, EndFrame : integer;
  end;


var
  frmSaveRange: TfrmSaveRange;


function DoSaveFrameRange(aFrameCount : integer): TSaveFrameRangeObject;


implementation


{$R *.dfm}


function DoSaveFrameRange(aFrameCount : integer): TSaveFrameRangeObject;
begin
  with TfrmSaveRange.Create(Application) do
    try
      FFrameCount := aFrameCount;

      Result.Process := False;

      ShowModal;

      if ModalResult = mrOK then begin
        Result.Process    := True;
        Result.StartFrame := StrToInt(eStartFrame.Text);
        Result.EndFrame   := StrToInt(eEndFrame.Text);
      end;

    finally
      Free;
    end;
end;


procedure TfrmSaveRange.FormCreate(Sender: TObject);
begin
  SetGUILanguageText;
end;


procedure TfrmSaveRange.SetGUILanguageText;
begin
  Caption := GLanguageHandler.Text[kSaveARangeOfFrames];

  lStart.Caption := GLanguageHandler.Text[kStart];
  lEnd.Caption := GLanguageHandler.Text[kEnd];

  bOK.Caption := GLanguageHandler.Text[kOK];
  bCancel.Caption := GLanguageHandler.Text[kCancel];
end;


procedure TfrmSaveRange.eStartFrameChange(Sender: TObject);
begin
  bOk.Enabled := ValidateInputs;
end;


function TfrmSaveRange.ValidateInputs: boolean;
var
  sf, ef : integer;

begin
  sf := StrToIntDef(eStartFrame.Text, -1);
  ef := StrToIntDef(eEndFrame.Text, -1);

  Result := (sf <> -1) and (ef <> -1) and (sf <= ef);
end;


end.

File: Source/Forms/FormSaveRange.cpp
// ===================================================================
//
//   (c) Paul Alan Freshney 2012-2024
//   www.freshney.org :: paul@freshney.org :: maximumoctopus.com
//
//   https://github.com/MaximumOctopus/LEDMatrixStudio
//
//   https://maximumoctopus.hashnode.dev/
//
//   C++ Rewrite October 11th 2023
//
// ===================================================================

#include <vcl.h>
#pragma hdrstop

#include "FormSaveRange.h"
#include "LanguageConstants.h"
#include "LanguageHandler.h"

#pragma package(smart_init)
#pragma resource "*.dfm"
TfrmSaveRange *frmSaveRange;

extern LanguageHandler *GLanguageHandler;

//---------------------------------------------------------------------------
//---------------------------------------------------------------------------

SaveFrameRangeObject OpenFrameRange(int frame_count)
{
	TfrmSaveRange *frmSaveRange = new TfrmSaveRange(Application);

	SaveFrameRangeObject sfto;

	frmSaveRange->MatrixFrameCount = frame_count;

	if (frmSaveRange->ShowModal() == mrOk)
	{
		sfto.Process = true;
		sfto.StartFrame = frmSaveRange->eStartFrame->Text.ToInt() - 1;
		sfto.EndFrame = frmSaveRange->eEndFrame->Text.ToInt() - 1;
	}

	delete frmSaveRange;

	return sfto;
}

//---------------------------------------------------------------------------
//---------------------------------------------------------------------------


__fastcall TfrmSaveRange::TfrmSaveRange(TComponent* Owner)
	: TForm(Owner)
{
	SetGuiLanguageText();
}


void __fastcall TfrmSaveRange::eStartFrameChange(TObject *Sender)
{
	bOK->Enabled = ValidateInputs();
}


void TfrmSaveRange::SetGuiLanguageText()
{
  Caption = GLanguageHandler->Text[kSaveARangeOfFrames].c_str();

  lStart->Caption = GLanguageHandler->Text[kStart].c_str();
  lEnd->Caption = GLanguageHandler->Text[kEnd].c_str();

  bOK->Caption = GLanguageHandler->Text[kOK].c_str();
  bCancel->Caption = GLanguageHandler->Text[kCancel].c_str();
}


bool TfrmSaveRange::ValidateInputs()
{
	int sf = eStartFrame->Text.ToIntDef(-1);
	int ef = eEndFrame->Text.ToIntDef(-1);

	return (sf >= 1 && ef >= 1 && sf <= ef && sf <= MatrixFrameCount && ef <= MatrixFrameCount);
}
